File: FE/mova/lib/app/models/favorite_model.dart
// To parse this JSON data, do
//
//     final favoriteModel = favoriteModelFromJson(jsonString);

import 'dart:convert';

FavoriteModel favoriteModelFromJson(String str) => FavoriteModel.fromJson(json.decode(str));

String favoriteModelToJson(FavoriteModel data) => json.encode(data.toJson());

class FavoriteModel {
  FavoriteModel({
    this.status,
    this.pesan,
    this.data,
  });

  int? status;
  String? pesan;
  List<Datum>? data;

  factory FavoriteModel.fromJson(Map<String, dynamic> json) => FavoriteModel(
    status: json["status"],
    pesan: json["pesan"],
    data: List<Datum>.from(json["data"].map((x) => Datum.fromJson(x))),
  );

  Map<String, dynamic> toJson() => {
    "status": status,
    "pesan": pesan,
    "data": List<dynamic>.from(data!.map((x) => x.toJson())),
  };
}

class Datum {
  Datum({
    this.id,
    this.filmId,
    this.userId,
    this.thumbnail,
    this.title,
    this.rating,
    this.category,
    this.createdAt,
    this.tipe,
    this.userUid,
  });

  int? id;
  int? filmId;
  int? userId;
  String? thumbnail;
  String? title;
  double? rating;
  String? category;
  DateTime? createdAt;
  String? tipe;
  String? userUid;

  factory Datum.fromJson(Map<String, dynamic> json) => Datum(
    id: json["id"],
    filmId: json["film_id"],
    userId: json["user_id"],
    thumbnail: json["thumbnail"],
    title: json["title"],
    rating: json["rating"].toDouble(),
    category: json["category"],
    createdAt: DateTime.parse(json["createdAt"]),
    tipe: json["tipe"],
    userUid: json["user_uid"],
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "film_id": filmId,
    "user_id": userId,
    "thumbnail": thumbnail,
    "title": title,
    "rating": rating,
    "category": category,
    "createdAt": createdAt!.toIso8601String(),
    "tipe": tipe,
    "user_uid": userUid,
  };
}

File: BE/mova/models/favorite_model.go
package models

import (
	"mova/db"
	"net/http"
)

func GetAllFavorites(user_id string) (Response, error) {
	var favorite Favorite
	arrFavorite := []Favorite{}
	var rating float32
	var res Response
	var err error

	con := db.CreateCon()

	sqlStatement1 := "SELECT * FROM favorite WHERE user_id = ($1) ORDER BY createdAt"
	sqlStatement2 := "SELECT AVG(rating) FROM rating WHERE film_id = ($1)"

	rows, err := con.Query(sqlStatement1, user_id)
	if err != nil {
		return res, err
	}
	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&favorite.Id, &favorite.FilmId, &favorite.UserId, &favorite.Thumbnail, &favorite.Title, &favorite.Rating, &favorite.Category, &favorite.CreatedAt, &favorite.Tipe, &favorite.UserUID)
		if err != nil {
			return res, err
		}

		err = con.QueryRow(sqlStatement2, favorite.FilmId).Scan(&rating)
		if err != nil {
			return res, err
		}
		favorite.Rating = rating

		arrFavorite = append(arrFavorite, favorite)
	}

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Get All Favorite"
	res.Data = arrFavorite

	return res, nil
}

func GetAllFavoritesFilter(user_id string, category string) (Response, error) {
	var favorite Favorite
	arrFavorite := []Favorite{}
	var rating float32
	var res Response
	var err error

	con := db.CreateCon()

	sqlStatement1 := "SELECT * FROM favorite WHERE user_id = ($1) AND category = ($2) ORDER BY createdAt"
	sqlStatement2 := "SELECT AVG(rating) FROM rating WHERE film_id = ($1)"

	rows, err := con.Query(sqlStatement1, user_id, category)
	if err != nil {
		return res, err
	}
	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&favorite.Id, &favorite.FilmId, &favorite.UserId, &favorite.Thumbnail, &favorite.Title, &favorite.Rating, &favorite.Category, &favorite.CreatedAt, &favorite.Tipe, &favorite.UserUID)
		if err != nil {
			return res, err
		}

		err = con.QueryRow(sqlStatement2, favorite.FilmId).Scan(&rating)
		if err != nil {
			return res, err
		}
		favorite.Rating = rating

		arrFavorite = append(arrFavorite, favorite)
	}

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Get All Favorite Filter"
	res.Data = arrFavorite

	return res, nil
}

func AddFavorite(film_id int, user_id int, thumbnail string, title string, rating float32, category string, createdAt string, tipe string, user_uid string) (Response, error) {
	var favorite Favorite
	var res Response
	var err error
	var id int

	con := db.CreateCon()

	sqlStatement := "INSERT INTO favorite (film_id, user_id, thumbnail, title, rating, category, createdAt, tipe, user_uid) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING id"

	err = con.QueryRow(sqlStatement, film_id, user_id, thumbnail, title, rating, category, createdAt, tipe, user_uid).Scan(&id)
	if err != nil {
		return res, err
	}

	favorite.Id = id
	favorite.FilmId = film_id
	favorite.UserId = user_id
	favorite.Thumbnail = thumbnail
	favorite.Title = title
	favorite.Rating = rating
	favorite.Category = category
	favorite.CreatedAt = createdAt
	favorite.Tipe = tipe
	favorite.UserUID = user_uid

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Add Favorite"
	res.Data = favorite

	return res, nil
}

func DeleteFavorite(id string) (Response, error) {
	var res Response
	var err error

	con := db.CreateCon()

	sqlStatement := "DELETE FROM favorite WHERE id = ($1)"

	_, err = con.Exec(sqlStatement, id)
	if err != nil {
		return res, err
	}

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Delete Favorite"
	res.Data = "Berhasil Delete Favorite"

	return res, nil
}
